### DISCUSSION TITLE
Welcome to spacy-hungarian-models Discussions!

### DISCUSSION THREAD

**Discussion by oroszgy** [huspacy/huspacy#21] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by amernokur** [comment]
Hi!

Sorry if these are too broad questions, you can disregard my comment if it doesn't fit the topic.

I'm developing - as a hobby non-profit project - a solution where for one use-case I need to calculate similarities between hungarian texts. At first I started using just a regular BagOfWords+cosine distance solution which was good enough initially to start things off. But now I'd like to improve the similarity calculation. Based on my initial research I've found the followings to be promising:  huBERT ( https://hlt.bme.hu/en/resources/hubert ), https://huggingface.co/NYTK/PULI-GPT-3SX, huspaCy (+NLKT?).
I've started fiddling around with huBERT and huspacy. My initial tests proved a tiny bit better at text similarity calculation than a simple bagofwords (*but I do not have a proper testing methodology yet, these are based on a small manual testing). However I think there is a lot of room for improvement (for example based on this: https://spacy.io/usage/spacy-101#vectors-similarity ).
What would you suggest (as general ideas) on how to improve my solution regarding hungarian text similarity problem? I'm also thinking of combining the bagofwords solution with the proposed new one to have a more reliable result. Also thinking of (and have been trying) to use tags (for most texts I have tags, but they are not usually of the same value set) or the result of NamedEntityRecognition (which at first glance is working pretty well in huspacy) but not sure how could I incorporate these the best. Sofar I've just appended it to the input text but as the model was trained on the wikipedia with full sentences this probably is a bit off. (my goal is sort of the give bigger weights to tags and/or named entities)
For the BoW solution I'm using word stemming (hunspell), but what I've gathered sofar I think it would be a bad practice to use stemming for huBERT or huSpacy. Can you confirm this?

(Note: I'd prefer to only use CPU and not GPU for calculation)

I know this is a lot, so I'd appreciate any kind of pointers! Feeling a bit lost in this huge field that has "improved" a lot in the last decade.

Thanks!

**Reply by oroszgy** [reply]
Hi!

Neither word vectors nor transformer models require stemming, so yes, this is not a good practice.

This approach worked well for me in the past: https://openreview.net/pdf?id=SyK00v5xx (https://github.com/PrincetonML/SIF) and it should be compatible with word embeddings shipped with the `lg` model. If you want something state-of-the-art, Sentence Transfomers might be a good idea to check out: https://www.sbert.net

Good luck with the project!

**Reply by amernokur** [reply]
Thank you very much! SBERT is proving to be fairly good (better) in my tests sofar. Haven't tried out your first suggestion, but definitely will.

On another subject: would you have any similar suggestions regarding categorization (text classification) of hungarian texts into a predetermined set of categories? I know this may be an even broader subject but was wondering if any of the above mentioned frameworks/libs and models could be used out of the box or not. (Based on my initial research it pretty much seems like I have to train my own model - which is understandable of course. I see that huspacy does not yet support textcategorizer ( https://spacy.io/api/doc#cats ). )
If I have to train my own model which "solution" would you suggest to a relative beginner?

Of course I understand if you prefer not to answer as it is not really related to huspacy. :)

**Reply by amernokur** [reply]
What I've tried sofar are:
- spacy with hu_core_news_lg model with a simple SVM training but ran into dimensions mismatch problem => still need to try with padding (not as easy as with BERT at first glance)
- spacy with hu_core_news_lg model: tried to add textcat pipeline but ran into lots of errors, maybe the model needs some other "stuff" as well? (last error: The pipeline could not be initialized because the vectors could not be found at 'models/hu_core_news_lg-parser-3.5.2/model-best'. If your pipeline was already initialized/trained before, call 'resume_training' instead of 'initialize', or initialize only the components that are new.)
- bert with SZTAKI-HLT/hubert-base-cc model using this https://huggingface.co/docs/transformers/tasks/sequence_classification modified so it is non-binary classification, but gave very poor results (probably I'm making some mistake somewhere or there is a problem with the model?)
- sbert with paraphrase-multilingual-MiniLM-L12-v2 model using simple SVM training: this is sofar the only one that produced a reasonable 0.7-0.8 accuracy (+also tried logistic regression as well, planning to try more types)

Just asking if these are good directions that I'm trying or I should try something else maybe?
If these kinds of posts should not be here you can delete and sorry.

*(by simple SVM I just mean this:
```
X_train, X_test, y_train, y_test = train_test_split(train_embeddings, train_labels, test_size=0.2, random_state=42)
classifier = SVC(kernel='linear')
classifier.fit(X_train, y_train)
y_pred = classifier.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)
```
)

**Reply by oroszgy** [reply]
Choosing the right text classification approach can heavily depend on the task. The number of labels, the size of the dataset, the domain of texts, the nature of task are important details that needs to be taken into consideration before building the model. 

In general:
- Tf-idf vectors with Logisitc Regression / SVM is a good baseline and could yield solid performance (see https://scikit-learn.org/stable/tutorial/text_analytics/working_with_text_data.html#training-a-classifier)
- HuBERT is often a good base model for various tasks, and one can easily use this model to build a classifier with transformers: https://huggingface.co/docs/transformers/tasks/sequence_classification
- Spacy also supports classification, but I recommend to generate a textcat config here and start with that: https://spacy.io/usage/training#quickstart

**Reply by amernokur** [reply]
Thank you very much!
